### PR TITLE
A good start for ImGuizmos

### PR DESCRIPTION
**Pull request by MrManiacc**
I've implemented the base/required functions to get the ImGuizmos library running. I plan to add support for things like ImGradients etc.

### CODE REVIEW

**Comment by MrManiacc** [comment]
Seems like all builds were fine other than the check style for some of the native functions. I don't think it's that big of a concern at the moment.

File: imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        ImGuizmo::Enable(&buffer[0]);
+    */
+
+    /**
+     * This will set the orthographic view
+     */
+    public static void setEnabled(boolean isEnabled) {
+        buffer[0] = isEnabled;
+        nEnabled(buffer);
+    }
```
Comment (SpaiR) [inline imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:32]: Very strange code. Just pass `boolean`, don't use array.

File: imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+
+    /**
+     * This will set the default window drawlist value
+     */
+    public static void setDrawList() {
+        setDrawList(ImGui.getWindowDrawList());
+    }
+
+    private native static void nBeginFrame(); /*
```
Comment (SpaiR) [inline imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:77]: You don't need this method. Just make `public static native void beginFrame` and call native function there. So it should be like `call native method` not `call java method->which calls native method`. Java layer is needed to do custom logic, like extraction of internal data from types `ImFloat` etc. Other than that, do direct native calls.

File: imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+
+    private native static void nDecomposeMatrixToComponents(float[] matrix, float[] translation, float[] rotation, float[] scale); /*
+        ImGuizmo::DecomposeMatrixToComponents(matrix, translation, rotation, scale);
+    */
+
+    /**
+     * This will allow for manual editing of the guizmos
+     */
+    public static void decomposeMatrixToComponents(float[] matrix, float[] translation, float[] rotation, float[] scale) {
+
```
Comment (SpaiR) [inline imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:96]: Wierd empty lines everywhere.

File: imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+        nSetRect(x, y, width, height);
+    }
+
+
+    /**
+     * This will set the rect position to the default current rect position
+     */
+    public static void setRect() {
+        ImVec2 pos = ImGui.getWindowPos();
+        ImVec2 size = ImGui.getWindowSize();
```
Comment (SpaiR) [inline imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:135]: Both of those methods create new objects. In binding I avoid unnecessary allocations as much as possible. You should do something like that https://github.com/SpaiR/imgui-java/blob/0ed891f391f70e1ba4edf863a6f8d22ccca23c98/imgui-binding/src/main/java/imgui/internal/ImGuiDockNode.java#L651-L660

File: imgui-binding/src/main/java/imgui/extension/imguizmo/Operation.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
+package imgui.extension.imguizmo;
+
+/**
+ * This is just the java representtation of the imguizmo operation
+ */
+public enum Operation {
```
Comment (SpaiR) [inline imgui-binding/src/main/java/imgui/extension/imguizmo/Operation.java:6]: Do not use enums. See any class in `flag` package to get the idea.

File: imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+    private native static void nRecomposeMatrixFromComponents(float[] translation, float[] rotation, float[] scale, float[] matrix); /*
+        ImGuizmo::RecomposeMatrixFromComponents(translation, rotation, scale, matrix);
+    */
+
+
+    /**
+     * This will allow for manual editing of the guizmos
+     */
+    public static void recomposeMatrixFromComponents(float[] matrix, float[] translation, float[] rotation, float[] scale) {
```
Comment (SpaiR) [inline imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:110]: Arguments order should be the same as for native method. Not something randomly generated. 😕

File: imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+
+    private native static void nDrawCubes(float[] view, float[] projection, float[] matrices, int matrixCount); /*
+        ImGuizmo::DrawCubes(view, projection, matrices, matrixCount);
+    */
+
+    /**
+     * This will allow us to draw an arbitrary cube in the world.
+     */
+    public static void drawCubes(float[] view, float[] projection, float[]... cubeMatrices) {
+        float[] matrices = new float[cubeMatrices.length * 16];
```
Comment (SpaiR) [inline imgui-binding/src/main/java/imgui/extension/imguizmo/ImGuizmo.java:159]: Speaking about unnecessary allocations, that is the case too.

**Review by SpaiR — CHANGES_REQUESTED** [review]
Great stuff! Thanks for your contribution, I very apprecciate your input.

Problems I see from the first look:
- Cleanups. `imgui-binding/tmp` should not exist. As well as shadow plugin etc. If you need to test out your code - use example module. It could be started with locally built libraries.
- Speaking about example: you need to provide one. As it's done for node-graphs bindings. It should demonstrotate how to use binding in Java ecosystem. Number of features you can show there depends on you, but it's required to exist for sure.
- It's `static native void`, not `native static void`. Please, keep code consistency.

Generally speaking, code is fine, but you need to pay more attention to details. And to revamp your code so it will follow style of other parts of the binding. I see that it's a draft implementation, so please mark your PR title in an appropriate way. Or make the PR itself as a draft (github has such functionality). Other than that, good job!

**Comment by AAstroPhysiCS** [comment]
Hi @MrManiacc,
I just wanted to ask if you're still working on this feature.
I am asking this because I plan to integrate this feature as well and in fact, I wanted to make a PR about it. Since, you've already done it, I did not want to make a another PR and just wait for you to finish.
If you stopped working on this feature or have limited time, I'll gladly take over your work and integrate this by myself.

Thanks for your contribution!

**Comment by MrManiacc** [comment]
> Hi @MrManiacc,
> I just wanted to ask if you're still working on this feature.
> I am asking this because I plan to integrate this feature as well and in fact, I wanted to make a PR about it. Since, you've already done it, I did not want to make a another PR and just wait for you to finish.
> If you stopped working on this feature or have limited time, I can gladly take over your work and integrate this by myself.
> 
> Thanks for your contribution!

Hi, feel free to take over. I'm going to be very busy for the next couple of weeks.

**Comment by SpaiR** [comment]
Implemented in #54

**Pull request closed without merging**